zerver/tests/test_auth_backends.py

```diff
@@ -1039,32 +1047,69 @@ def stage_two_of_registration(self, result: HttpResponse, realm: Realm, subdomai
                                  parsed_url.path)
         self.assertTrue(uri.startswith('http://zulip.testserver/accounts/login/subdomain/'))
 
-        result = self.client_get(result.url)
+        if not user_avatar_url:
+            result = self.client_get(result.url)
+        else:
+            data['user_avatar_url'] = user_avatar_url
+            data['next'] = ''
+            token = store_login_data(data)
+            url_path = reverse('zerver.views.auth.log_into_subdomain', args=[token])
+            result = self.client_get(url_path, subdomain=subdomain)
 
         self.assertEqual(result.status_code, 302)
         confirmation = Confirmation.objects.all().last()
         confirmation_key = confirmation.confirmation_key
         self.assertIn('do_confirm/' + confirmation_key, result.url)
+        if user_avatar_url:
+            # If there is an avatar_url, then there must be a user_avatar_url
+            # on the preregistration user.
+            prereg_user = confirmation.content_object
+            self.assertIsNotNone(prereg_user.user_avatar_url)
+
         result = self.client_get(result.url)
+
         self.assert_in_response('action="/accounts/register/"', result)
-        data = {"from_confirmation": "1",
-                "key": confirmation_key}
+        if user_avatar_url:
+            data = {"from_confirmation": "1",
+                    "user_avatar_url": user_avatar_url,
```

Comment:
Why is `user_avatar_url` being POSTed to /accounts/register? :thinking: I think in general this is the point where I stop agreeing the implementation of the test. We should be emulating the real flow that the browser will do. So, all that should be POSTed is the usual dict
```
            data = {"from_confirmation": "1",
                    "key": confirmation_key}
}
```